### PR TITLE
Extend validator CLI tool to support multisig.

### LINKED ISSUE #269: Write cli tool to update validator set
## Use cases
* As "partner manager"? I want to add a new validator to the active set of validators in order to create a more resilient network.
* As "partner manager"? I want to remove a validator from the active set of validators in order to punish malicious behaviour and curate the network.

## What to do
- [x] Build a CLI tool (in weave) that can be used in hugnet to **add** a new validator.

Steps internally are: 
* query current validator set
* read private key from stdIn or ENV (base64 encoded) (a file would require encryption/ decryption)
* build, sign and send a [SetValidatorsMsg](https://github.com/iov-one/weave/blob/master/x/validators/codec.proto#L21) to the cluster

- [x] Analogue to this we want to **remove** an validator.
* Remove is be done by setting the validator `Power` to `0`

- [x] Extend the CLI tool to support **multisig** contracts
* This includes exporting the generated message into a file instead of sending it to the cluster
* Importing a message file to add a signature and storing it
* Sending a stored message to the cluster

## Technical background
* The [ABCI spec](https://tendermint.com/docs/app-dev/abci-spec.html#request-response-messages) provides a response field to curate the active validator set.
* The initial validator set is defined in the [genesis file](https://tendermint.com/docs/tendermint-core/using-tendermint.html#initialize)
* Roman had created a [`x/validators`(https://github.com/iov-one/weave/tree/master/x/validators) package with an `initializer` to read and persist them
* An [update handler](https://github.com/iov-one/weave/blob/master/x/validators/handler.go#L40) and [query handler](https://github.com/iov-one/weave/blob/master/x/validators/handler.go#L27) exist
* The query handler should return the accounts that are allowed to sign an update message
* The current validator set can be queried from tendermint via plain http `/validators`

## Not in Scope
* There is some potential risk for a concurrent modification as we are not protecting the data via version number, timestamp, etc. We could also split the update into separate add/ remove operations to prevent overwrites. If we want to address this, any solution should be done in a separate story and after this one.

### PR DESCRIPTION
**Pull request by husio**
An example flow:

1. Create a transaction that configures a validator. This transaction
points to a multisig contract as an authentication method:
```
    ./validators multisig-new \
        -power 7 \
	-pubkey HsEV3BeuXBHWmpdLkdk8R43kGzZYeVhUspqrIYacJUw= \
	-multisig 5AE2C58796B0AD48FFE7602EAC3353488C859A2B \
	> request.tx
```
2. Send the `request.tx` around and ask people to sign it:
```
    ./validators multisig-sign \
        -tm http://127.0.0.1:22345 \
	-key d34c1970ae90acf3405f2d99dcaca16d0c7db379f4beafcfdf667b9d69ce350d27f5fb440509dfa79ec883a0510bc9a9614c3d44188881f0c5e402898b4bf3c9 \
	< request.tx \
	> request_signed_1.tx
```
3. Once enough signatures are collected, submit the transaction:
```
    ./validators multisig-submit \
        -tm http://127.0.0.1:22345 \
	< request.signed_1.tx
```
Multisig addresses: https://github.com/iov-one/ops/wiki/MulitiSig-contract-addresses

resolve #269

### CODE REVIEW

File: cmd/validators/main.go [hunk 1]
```diff
@@ -59,26 +60,32 @@ Available commands: %s
...
-	if err != nil {
-		return fmt.Errorf("cannot list validators: %s\n", err)
-	}
-	b, err := json.MarshalIndent(info, "", "\t")
-	if err != nil {
-		return fmt.Errorf("cannot serialize: %s", err)
+		var (
+			tmAddrFl = flag.String("tm", "https://bns.NETWORK.iov.one:443", "Tendermint node address. Use proper NETWORK name.")
+		)
+		flag.Parse()
```
Comment (alpe) [inline cmd/validators/main.go]: req: please replace the default with a new `FlagSet` to get rid of all the test flag noise.
https://stackoverflow.com/questions/42243919/using-go-build-but-i-also-see-the-test-flags

File: cmd/validators/main.go [hunk 2]
```diff
@@ -120,12 +127,159 @@ type pubKeyInfo struct {
...
+	return func() error {
+		flag.Usage = func() {
+			fmt.Fprint(flag.CommandLine.Output(), `
+Create a new validator update request with multi signature authentication.
+Created request is binary serialized and written to standard output.
+
+Returned request must be signed by other parties before it can be submitted.
+
+`)
+			flag.PrintDefaults()
```
Comment (alpe) [inline cmd/validators/main.go]: req: same as above to use a new `FlagSet`. Please also check the other calls.

File: cmd/validators/main.go [hunk 3]
```diff
@@ -120,12 +127,159 @@ type pubKeyInfo struct {
...
+				},
+				Power: *powerFl,
+			},
+		)
+
+		raw, err := addValidatorTx.Marshal()
+		if err != nil {
+			return fmt.Errorf("cannot serialize transaction: %s", err)
+		}
+		out.Write(raw)
```
Comment (alpe) [inline cmd/validators/main.go]: pp: you could return the err instead of ignoring it.

File: cmd/validators/main.go [hunk 4]
```diff
@@ -120,12 +127,159 @@ type pubKeyInfo struct {
...
+		if seq, err := aNonce.Next(); err != nil {
+			return fmt.Errorf("cannot get the next sequence number: %s", err)
+		} else {
+			client.SignTx(&tx, key, genesis.ChainID, seq)
+		}
+
+		if raw, err := tx.Marshal(); err != nil {
+			return fmt.Errorf("cannot serialize transaction: %s", err)
+		} else {
+			out.Write(raw)
```
Comment (alpe) [inline cmd/validators/main.go]: pp: please return the err

File: cmd/validators/main.go [hunk 5]
```diff
@@ -120,12 +127,159 @@ type pubKeyInfo struct {
...
+			return errors.New("no stdin data")
+		}
+		var tx app.Tx
+		if err := tx.Unmarshal(raw); err != nil {
+			return fmt.Errorf("cannot deserialize transaction: %s", err)
+		}
+		bnsClient := client.NewClient(client.NewHTTPConnection(*tmAddrFl))
+
+		if err := bnsClient.BroadcastTx(&tx).IsError(); err != nil {
+			return fmt.Errorf("cannot broadcast transaction: %s", err)
```
Comment (alpe) [inline cmd/validators/main.go]: pp: no need for the `if`, just return the err

File: cmd/validators/main.go [hunk 6]
```diff
@@ -120,12 +127,159 @@ type pubKeyInfo struct {
...
+	}
+}
+
 func cmdAdd() error {
 	var (
 		tmAddrFl = flag.String("tm", "https://bns.NETWORK.iov.one:443", "Tendermint node address. Use proper NETWORK name.")
 		pubKeyFl = flag.String("pubkey", "", "Base64 encoded, ed25519 public key.")
 		hexKeyFl = flag.String("key", "", "Hex encoded, private key of the validator that is to be added/updated.")
-		powerFl  = flag.Int64("power", 10, "Validator node power. Set to 0 to delete a node.")
+		powerFl  = flag.Int64("power", 2, "Validator node power. Set to 0 to delete a node.")
```
Comment (alpe) [inline cmd/validators/main.go]: q: why 2? currently 10 is default in the testnets? The use should enter the value rather than us setting a default value, IMHO.

File: cmd/validators/main.go [hunk 7]
```diff
@@ -59,26 +60,32 @@ Available commands: %s
...
+
+		info, err := listValidators(*tmAddrFl)
+		if err != nil {
+			return fmt.Errorf("cannot list validators: %s\n", err)
+		}
+		b, err := json.MarshalIndent(info, "", "\t")
+		if err != nil {
+			return fmt.Errorf("cannot serialize: %s", err)
+		}
+		out.Write(b)
```
Comment (alpe) [inline cmd/validators/main.go]: pp: return out.Write error

File: cmd/validators/main.go (hunk 2 above)
Comment (husio) [reply cmd/validators/main.go]: This is a workaround, but I like it!

I wish we could locate what is polluting our flags and fix it there.

File: cmd/validators/main.go (hunk 5 above)
Comment (husio) [reply cmd/validators/main.go]: I need the if because I am wrapping the error. :thinking:

File: cmd/validators/main.go (hunk 6 above)
Comment (husio) [reply cmd/validators/main.go]: :man_shrugging: Changing back to 10.

**Comment by husio** [comment]
@alpe please have a look again.

File: cmd/validators/main.go (hunk 5 above)
Comment (ethanfrey) [reply cmd/validators/main.go]: Maybe this can be done easier with the new error package

Comment (husio) [reply cmd/validators/main.go]: I am wondering if I cannot see something obvious here. Let me explain what I think and tell me if it makes sense.

You are both talking about below code:
```go
if err := bnsClient.BroadcastTx(&tx).IsError(); err != nil {
    return fmt.Errorf("cannot broadcast transaction: %s", err)
}
```
The above could be replaced with a shorter version:
```go
return bnsClient.BroadcastTx(&tx).IsError()
```

The second example is correct, but the returned error does not provide the same information as the in the first example. I think wrapping an error with a piece of additional information is helpful.

If the new `errors` package is merged, `fmt.Errorf` could be replaced with
```go
if err := bnsClient.BroadcastTx(&tx).IsError(); err != nil {
    return errors.Wrap(err, "cannot broadcast transaction")
}
```
But the `if` must stay. Otherwise a new error would be created, even if `IsError` call returns `nil`.

Does this make sense?

Comment (ethanfrey) [reply cmd/validators/main.go]: @husio Yes, with the current API. But in the refactor, you removed a nice feature of the "legacy" errors API. Look at https://github.com/iov-one/weave/pull/308#discussion_r255426591

```golang
// Wrap safely takes any error and promotes it to a TMError.
// Doing nothing on nil or an incoming TMError.
func Wrap(err error) TMError {
	// nil or TMError are no-ops
	if err == nil {
		return nil
	}
```

Comment (ethanfrey) [reply cmd/validators/main.go]: You do not need to maintain "Do nothing on TmError", but at least "return nil on nil" is a nice feature

**Comment by ethanfrey** [comment]
Code looks good, and I like the direction. Two points to add though:

1. When I create a multisig transaction, I know what it is, but there is no clear way to view `request.tx`, and thus all signers are basically blindly signing. Can you add a `multisig-view` command to show the contents of such a (possibly signed) request?

2. I see no tests... Maybe not a proper CI unit test (very hard with cli tools), but how about a bash script that runs through the happy path against a local instance? Or maybe you have a better idea. Anything is fine... I just didn't feel like running it manually and setting up the chain, so a quick automatic check would be nice.

File: cmd/validators/main.go (hunk 5 above)
Comment (husio) [reply cmd/validators/main.go]: I understand now. I was not aware of this functionality and I also did not carry it to the new `Wrap`. I is an easy change to add it back.

_tl;dr;_ I am afraid that adding support for `Wrap(<nil>, "message")` may cause subtle bugs. But I can apply this change if this is the pattern we want to follow in `weave`.

I understand that supporting `nil` error in `Wrap` function as proposed may look appealing. It allows saving on an `if` statement when last operation done in a function can return an error. It is easier to explain with an example:

```go
func foo() error {
   if err := first(); err != nil {
       return errors.Wrap(err, "first")
   }
   return errors.Wrap(second(), "second")
}
```
When calling `second`, I can wrap the result without checking if the error is `nil`. When calling `first`, I have to test for an error anyway, because the result affects the flow.

Consider the verbose version of the very same functionality:
```go
func foo() error {
   if err := first(); err != nil {
       return errors.Wrap(err, "first")
   }
   if err := second(); err != nil {
       return errors.Wrap(err, "second")
   }
   return nil
}
```
For me it is clear what the flow is. While `errors.Wrap(second(), "second")` is very dense and expands to several operations, second example stays very simple.

I would not be surprised if most programmers asked to explain the above two examples were often wrong with the first one and always right with the second one.

Maybe it all boils down to a personal preference of clarity over expressiveness. I always tend to pick the verbose approach thinking that future Piotr will appriciate me not trying to be too smart :wink:

Comment (ethanfrey) [reply cmd/validators/main.go]: It is not about the first or the second usage everywhere.
It is about the ability to choose.

I think there are cases when the second usage is cleaner. When there is a long series of checks, as you mentioned. 

I think there are also cases when the first usage is cleaner. When there is only one error return, and the overhead of needing the if statement just to wrap it will cause programmer to NOT wrap it out of laziness.

I want to make it easy for those cases, and allow the developers using the errors package to choose. I don't think there are many cases where I relied on the `nil -> nil` functionality of the legacy `Wrap()`, but definitely a few places

Comment (ethanfrey) [reply cmd/validators/main.go]: Also, another example to your above...

Common (unwrapped) idiom - I think putting the logic in an if statement makes it harder to read as well.

```golang
func foo() error {
   err := first()
   if err != nil {
       return err
   }
   err = second()
   return errors.Wrap(err, "second")
```

And a wrapped version

```golang
func foo() error {
   err := first()
   if err != nil {
      return errors.Wrap(err, "first")
   }
   err = second()
   return errors.Wrap(err, "second")
}
```

**Comment by husio** [comment]
@ethanfrey please have a look at the last two commits and let me know if this is what you had in mind.
I have added a `mutisig-view` command and a test that goes through the multisig flow.

File: cmd/validators/main_test.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+									"admin_threshold": 2,
+									"activation_threshold": 1
+								}
+								]
+							}
+						}
+					}
+				}
+			`)
+		case r.Method == "POST" && r.URL.Path == "/":
```
Comment (ethanfrey) [inline cmd/validators/main_test.go]: I think there are many paths there... like `/broadcast_tx_commit`.
Also, don't you read the nonce from the chain?

File: cmd/validators/main.go (hunk 5 above)
Comment (husio) [reply cmd/validators/main.go]: I have pushed the changes required for this https://github.com/iov-one/weave/pull/275/commits/9d1614508ab365f5638be151f40eccc58fa37198
Once new errors is merged, I can update this line.

File: cmd/validators/main_test.go (hunk 8 above)
Comment (ethanfrey) [reply cmd/validators/main_test.go]: Maybe there is an easier way than mocking it all?
If not, then it is quite a bit of work to build a proper mock that will return nice results, and I would mock out on the bnsClient level, which has higher-level API

Comment (husio) [reply cmd/validators/main_test.go]: Test that is provided is using only `POST /`.

I was aiming for the minimal server implementation that will allow passing the tests. I do not understand the tendermint protocol enough to try to provide a solid implementation.

I think the nonce is read from the chain with a fallback if it does not exist. I think this is also done with `POST /` because I did not log another request.

Below is the full test run output. Lines with `tendermint request:` are important:
```
$ go test github.com/iov-one/weave/cmd/validators/... -logrequest -v
2019/02/11 13:32:42 proto: duplicate proto type registered: common.KVPair
2019/02/11 13:32:42 proto: duplicate proto type registered: common.KI64Pair
=== RUN   TestMultisig
--- PASS: TestMultisig (0.00s)
    main_test.go:34: unsigned request: {
                "Sum": {
                        "SetValidatorsMsg": {
                                "validator_updates": [
                                        {
                                                "pubkey": {
                                                        "type": "ed25519",
                                                        "data": "j4JRVstX"
                                                },
                                                "power": 7
                                        }
                                ]
                        }
                }
        }
    main_test.go:68: tendermint request: GET /genesis: 
    main_test.go:68: tendermint request: POST /: {"jsonrpc":"2.0","id":"jsonrpc-client","method":"abci_query","params":{"data":"E28AE9A6EB94FC88B73EB7CBD6B87BF93EB9BEF0","height":"0","path":"/auth","prove":false}}
    main_test.go:50: signed request: {
                "signatures": [
                        {
                                "pubkey": {
                                        "Pub": {
                                                "Ed25519": "J/X7RAUJ36eeyIOgUQvJqWFMPUQYiIHwxeQCiYtL88k="
                                        }
                                },
                                "signature": {
                                        "Sig": {
                                                "Ed25519": "6EJlCaTdYE/dCptm7Jq/6drRS/e95iKTBaNfPcgf9k53VFiqsuXBPwGYPC7Htn1rK+tyYnVw/xr9W4Wy1tjvCA=="
                                        }
                                }
                        }
                ],
                "Sum": {
                        "SetValidatorsMsg": {
                                "validator_updates": [
                                        {
                                                "pubkey": {
                                                        "type": "ed25519",
                                                        "data": "j4JRVstX"
                                                },
                                                "power": 7
                                        }
                                ]
                        }
                }
        }
    main_test.go:68: tendermint request: POST /: {"jsonrpc":"2.0","id":"jsonrpc-client","method":"broadcast_tx_commit","params":{"tx":"EmgSIgogJ/X7RAUJ36eeyIOgUQvJqWFMPUQYiIHwxeQCiYtL88kiQgpA6EJlCaTdYE/dCptm7Jq/6drRS/e95iKTBaNfPcgf9k53VFiqsuXBPwGYPC7Htn1rK+tyYnVw/xr9W4Wy1tjvCNIDFwoVChEKB2VkMjU1MTkSBo+CUVbLVxAH"}}
PASS
ok      github.com/iov-one/weave/cmd/validators 0.021s
```

Comment (ethanfrey) [reply cmd/validators/main_test.go]: Good call, yes
> POST /: {"jsonrpc":"2.0","id":"jsonrpc-client","method":"abci_query","params":{"data":"E28AE9A6EB94FC88B73EB7CBD6B87BF93EB9BEF0","height":"0","path":"/auth","prove":false}}
is the line I was thinking of, didn't realize which protocol uses which path (there is URI/HTTP GET, JsonRPC/HTTP POST, and Websockets).

Comment (ethanfrey) [reply cmd/validators/main_test.go]: I will do one more pass and then give you a :+1:

**Comment by husio** [comment]
Unless there is anything more I should do, I will merge this in 1-2 hours.
